tidy3d/components/source.py

```
@@ -509,6 +500,101 @@ def _warn_if_large_number_of_freqs(cls, val):
+        fmin, fmax = source_time.frequency_range()
+        for name, scalar_field in val.field_components.items():
+            freqs = scalar_field.f
+            if len(freqs) != 1:
+                raise SetupError(
+                    f"`field_data.{name}` must have a single frequency, "
+                    f"contains {len(freqs)} frequencies."
+                )
+            freq = float(freqs[0])
+            if (freq < fmin) or (freq > fmax):
+                raise SetupError(
+                    f"`field_data.{name}` contains frequency: {freq:.2e} Hz, which is outside "
+                    f"of the source's `source_time` frequency range [{fmin:.2e}-{fmax:.2e}] Hz."
+                )
+        return val
+
+    @pydantic.validator("field_dataset", always=True)
+    def _tangential_component_defined(cls, val: FieldDataset, values: dict) -> FieldDataset:
+        """Assert that at least one tangential field component is provided."""
+        size = get_value(key="size", values=values)
+        normal_axis = size.index(0.0)
+        _, (cmp1, cmp2) = cls.pop_axis("xyz", axis=normal_axis)
+        for field in "EH":
+            for cmp_name in (cmp1, cmp2):
+                tangential_field = field + cmp_name
+                if tangential_field in val.field_components:
+                    return val
+        raise SetupError("no tangential field found in the suppled `field_data`.")
+
+    @pydantic.validator("field_dataset", always=True)
+    def _tangential_fields_span_source(cls, val: FieldDataset, values: dict) -> FieldDataset:
+        """Assert that provided data spans source bounds in the frame with the source center as the
+        origin."""
+        size = get_value(key="size", values=values)
+        for field_name, field in val.field_components.items():
+            try:
+                for dim, dim_name in enumerate("xyz"):
+                    assert np.amin(field.coords[dim_name]) <= -size[dim] / 2 + DATA_SPAN_TOL
```

I would just raise the setup error directly here:
```py
for dim, dim_name in enumerate("xyz"):
    in_bounds_min = np.amin(field.coords[dim_name]) <= -size[dim] / 2 + DATA_SPAN_TOL
    in_bounds_max = np.amax(field.coords[dim_name]) >= size[dim] / 2 - DATA_SPAN_TOL
    if not (in_bounds_min and in_bounds_max):
         raise SetupError(f"Data for field {field_name} does not span the source plane.")